telemetry.go

```diff
@@ -0,0 +1,86 @@
+// Copyright (c) 2021 Snowflake Computing Inc. All right reserved.
+
+package gosnowflake
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"sync"
+)
+
+const (
+	telemetryPath    = "/telemetry/send"
+	defaultFlushSize = 100
+)
+
+type telemetryMessage struct {
+	Type    string `json:"type,omitempty"`
+	QueryID string `json:"QueryID,omitempty"`
+}
+
+type telemetryData struct {
+	Timestamp int64             `json:"timestamp,omitempty"`
+	Message   *telemetryMessage `json:"message,omitempty"`
+}
+
+type snowflakeTelemetry struct {
+	logs      []*telemetryData
+	flushSize int
+	sr        *snowflakeRestful
+	mutex     *sync.Mutex
+	enabled   bool
+}
+
+func (st *snowflakeTelemetry) addLog(data *telemetryData) error {
+	if !st.enabled {
+		return fmt.Errorf("telemetry disabled; not adding log")
+	}
+	st.mutex.Lock()
+	defer st.mutex.Unlock()
+	st.logs = append(st.logs, data)
+	if len(st.logs) >= st.flushSize {
+		if err := st.sendBatch(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (st *snowflakeTelemetry) sendBatch() error {
+	if !st.enabled {
+		return fmt.Errorf("telemetry disabled; not sending log")
+	}
+	type telemetry struct {
+		Logs []*telemetryData `json:"logs"`
+	}
+	s := &telemetry{st.logs}
```

Comment:
sendBatch can be called from outside and not just from addLog, so there should be a lock protection on st.log. Otherwise there might be some logs added but get lost since at the end of this sendBatch, st.logs is cleaned. 

Should we consider doing a sendBatch when the connection is closed? Otherwise we are expecting other developers to take care when to sendBatch.